addons/mail/models/mail_message.py

```diff
@@ -848,12 +857,145 @@ def _notify(self, force_send=False, send_after_commit=True, user_signature=True)
             ])._notify(self, force_send=force_send, send_after_commit=send_after_commit, user_signature=user_signature)
 
         notif_partners._notify_by_chat(self)
-
         channels_sudo._notify(self)
-
         # Discard cache, because child / parent allow reading and therefore
         # change access rights.
         if self.parent_id:
             self.parent_id.invalidate_cache()
 
         return True
+
+    @api.multi
+    def moderate(self, decision):
+        """ :param decision
+                 * accept       - moderate message and broadcast that message to followers of relevant channels.
+                 * reject       - message will be deleted from the database without broadcast
+                                  an email sent to the author with an explanation that the moderators can edit.
+                 * discard      - message will be deleted from the database without broadcast.
+                 * allow        - add email address to white list people of specific channel,
+                                  so that next time if a message come from same email address on same channel,
+                                  it will be automatically broadcasted to relevant channels without any approval from moderator.
+                 * ban          - add email address to black list of emails for the specific channel.
+                                  From next time, a person sending a message using that email address will not need moderation.
+                                  message_post will not create messages with the corresponding expeditor.
+        """
+        # For now, his method is called on a singleton when decision is 'ban' or 'allow'
+        if decision == 'accept':
+            self._accept_message()
+        elif decision in ['reject', 'discard']:
+            self._notify_deletion_and_unlink()
+        else:
+            channel = self.env['mail.channel'].browse(self.res_id)
+            channel._add_to_moderation_email([self.email_from], decision)
+            self._accept_message(True) if decision == 'allow' else self.notify_deletion_and_unlink(True)
+        return True
+
+    def _accept_message(self, allow=False):
```

Comment:
This method mixes two things, updating status of messages and notifying people, and searching for other messages. I don't really understand the code flow. I think the input / output are strange.

Reply:
Now the method does not search for other messages but it is still notifies and changes status of messages.